control/discovery.py

```diff
@@ -0,0 +1,1077 @@
+            struct.unpack_from('<256B', data, 584)
+
+        pdu_reply = Pdu()
+        pdu_reply.type = NVME_TCP_PDU.RSP
+        pdu_reply.header_length = 24
+        pdu_reply.packet_length = 24
+
+        connect_reply = CqeConnect()
+        connect_reply.controller_id = int(self_conn.controller_id)
+        connect_reply.sq_head_ptr = self_conn.sq_head_ptr
+        connect_reply.cmd_id = cmd_id
+
+        try:
+            conn.sendall(pdu_reply + connect_reply)
+        except BrokenPipeError:
+            self.logger.error("client disconnected unexpectedly.")
+            return -1
+        self.logger.debug("reply connect request.")
+        return 0
+
+    def reply_fc_cmd_prop_get(self, conn, data, cmd_id):
+        """Reply property get request."""
+
+        self.logger.debug("handle property get request.")
+        self_conn = self.conn_vals[conn.fileno()]
+        nvmeof_prop_get_set_rsvd0 = struct.unpack_from('<35B', data, 13)
+        # property size = (attrib+1)x4, 0x1 means 8 bytes
+        nvmeof_prop_get_set_attrib = struct.unpack_from('<1B', data, 48)[0]
+        nvmeof_prop_get_set_rsvd1 = struct.unpack_from('<3B', data, 49)
+        nvmeof_prop_get_set_offset = struct.unpack_from('<I', data, 52)[0]
+
+        pdu_reply = Pdu()
+        pdu_reply.type = NVME_TCP_PDU.RSP
+        pdu_reply.header_length = 24
+        pdu_reply.packet_length = 24
+
+        # reply different property data
+        property_get = CqePropertyGetSet()
+        if NVME_CTL(nvmeof_prop_get_set_offset) == NVME_CTL.CAPABILITIES:
+            # controller capabilities
+            # \x7f = maxinum queue entries support:128
+            # \x01 contiguous queues required: true
+            # \x1e timeout(to ready status): 1e(15000ms default in server side)
+            # \x20 command sets supportd: 1 (NVM IO command set)?
+            property_get.property_data = (c_ubyte * 8)(0x7f, 0x00, \
+                0x01, 0x1e, 0x20, 0x00, 0x00, 0x00)
+        elif NVME_CTL(nvmeof_prop_get_set_offset) == NVME_CTL.CONFIGURATION:
+            # b'\x00\x00\x46\x00\x00\x00\x00\x00'
+            # 0x46: IO Submission Queue Entry Size: 0x6 (64 bytes)
+            # IO Completion Queue Entry Size: 0x4 (16 bytes)
+            property_data = (c_ubyte * 8)(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
+            if len(self_conn.property_data) == 8:
+                property_data = (c_ubyte * 8)(*self_conn.property_data)
+            property_get.property_data = property_data
+        elif NVME_CTL(nvmeof_prop_get_set_offset) == NVME_CTL.STATUS:
+            property_data = (c_ubyte * 8)(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
+            if len(self_conn.property_data) == 8:
+                property_data = (c_ubyte * 8)(*self_conn.property_data)
+            shutdown_notification = ((property_data[1]) >> 6) & 0x3
+            if shutdown_notification == 0:
+                if self_conn.property_set is True:
```

Comment:
The condition should check value of CC.EN

Reply:
I initially wrote it as `str: property_data` preallocate 8 bytes initialized to zero. But there are always conversion errors such as `str`, `tuple`, and `cbyte`, and I have already written this process here https://github.com/ceph/ceph-nvmeof/pull/127#issuecomment-1696943699. Forced, I used judgment methods.  

When the value set by the previous message is not stored in advance, it is in a state of all zeros, and it will enter the else branch, returning:
```
                else:
                    property_get.property_data = (c_ubyte * 8)(0x00, 0x00, \
                        0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
```
So logically, I don't think there's any problem. There may be better ways to convert `property_data` types .